File: mbfpp/code/ScalaExamples/src/main/scala/com/mblinn/mbfpp/oo/visitor/Examples.scala
package com.mblinn.mbfpp.oo.visitor

object Examples {
	trait Person {
	  def fullName: String
	  def firstName: String
	  def lastName: String
	  def houseNum: Int
	  def street: String
	}
	
	class SimplePerson(val firstName: String, val lastName: String,
			val houseNum: Int, val street: String) extends Person {
	  def fullName = firstName + " " + lastName
	}
	
	class ComplexPerson(name: Name, address: Address) extends Person {
	  def fullName = name.firstName + " " + name.lastName
	  
	  def firstName = name.firstName
	  def lastName = name.lastName
	  def houseNum = address.houseNum
	  def street = address.street
	}
	class Address(val houseNum: Int, val street: String)
	class Name(val firstName: String, val lastName: String)
	
	implicit class ExtendedPerson(person: Person) {
	  def fullAddress = person.houseNum + " " + person.street
	}
}

File: mbfpp/code/ClojureExamples/src/mbfpp/oo/visitor/examples.clj
(ns mbfpp.oo.visitor.examples)

(defprotocol NameExtractor
  (extract-name [this] "Extracts a name from a person."))

(defrecord SimplePerson [first-name last-name house-num street])

(extend-type SimplePerson
  NameExtractor
  (extract-name [this]
    (str (:first-name this) " " (:last-name this))))

(defrecord ComplexPerson [name address]
  NameExtractor
  (extract-name [this]
    (str (-> this :name :first) " " (-> this :name :last))))

(defprotocol
  AddressExtractor
  (extract-address [this] "Extracts and address from a person."))

(extend-type SimplePerson
  AddressExtractor
  (extract-address [this]
                   (str (:house-num this) " " (:street this))))

(extend-type ComplexPerson
  AddressExtractor
  (extract-address [this]
                   (str (-> this :address :house-num)
                        " "
                        (-> this :address :street))))

(defmulti test-multimethod (fn [keyword] keyword))

(defmethod test-multimethod :foo [a-map]
  "foo-method was called")
  
(defmethod test-multimethod :bar [a-map]
  "bar-method was called")

(defmulti perimeter (fn [shape] (:shape-name shape)))
(defmethod perimeter :circle [circle]
  (* 2 Math/PI (:radius circle)))
(defmethod perimeter :rectangle [rectangle]
  (+ (* 2 (:width rectangle)) (* 2 (:height rectangle))))

(def some-shapes [{:shape-name :circle :radius 4}
                   {:shape-name :rectangle :width 2 :height 2}])

(defmulti area (fn [shape] (:shape-name shape)))
(defmethod area :circle [circle]
  (* Math/PI (:radius circle) (:radius circle)))
(defmethod area :rectangle [rectangle]
  (* (:width rectangle) (:height rectangle)))

(defmethod perimeter :square [square]
  (* 4 (:side square)))
(defmethod area :square [square]
  (* (:side square) (:side square)))

(def more-shapes (conj some-shapes
                       {:shape-name :square :side 4}))
  


  


